File: src/main/kotlin/easy/PalindromeNumber.kt
package easy

class PalindromeNumber {
    /*
     * Link: https://leetcode.com/problems/palindrome-number/
     * Difficult: Easy
     */
    fun isPalindrome(x: Int): Boolean {
        var reversedNumber: Int = 0
        var number = x

        if(x<0 || x % 10 == 0 && x != 0){
            return false
        }

        while (number > reversedNumber) {
            reversedNumber = reversedNumber * 10 + number % 10
            number /= 10
        }

        return reversedNumber == number || number == reversedNumber/10
    }
}

File: leetcode/easy/palindrome_number.cpp
#include "../data_structures.h"
#include <bits/stdc++.h>

/*
 * Number: 09
 * Difficult: Easy
 * Link: https://leetcode.com/problems/palindrome-number
 * Tags: Math
 */ 

class PalindromeNumber {
public:
    /*
     * Time complexity: O(N)
     * Space complexity: O(N)
     * 
     * Time beats 14.16%
     * Memory beats 39.25%
     */
    bool isPalindrome(int x) {
        std::string s = std::to_string(x);
        int left = 0;
        int right = s.size() - 1;

        for(;;) {
            if(left >= right) {
                break; 
            }

            if(s[left] != s[right]) {
                return false;
            }

            left++;
            right--;
        }

        return true;
    }

    // Improve space complexity to O(1)
};